File: Knapsack/Repeated/knapsack.py
"""
    ****************DYNAMIC PROGRAMMING******************
    * This program computes discrete knapsack value with repetitions allowed
    * It stores optimized  discrete knapsack value for each weight 
    * from [1.....W] where W is the total_capacity of the given knapsack.
    * It is same as Coin change problem (Dynamic programming).
    * It takes O(n * W) time.
"""

class item:
    def __init__(self, weight, value):
        self.weight = weight
        self.value = value

def knapsack(items, total_weight, total_items):
    values_of_weights = [0 for x in range(0, total_weight + 1)]
    current_value = 0
    for i in range(1, total_weight + 1):
        for j in range(0 , total_items):
            if items[j].weight <= i:
                current_value = values_of_weights[i - items[j].weight] + items[j].value;
            if current_value > values_of_weights[i]:
                values_of_weights[i] = current_value
    
    return values_of_weights[total_weight]

def main():
    total_weight = int(input("Give total capacity of knapsack :"))
    total_items = int(input("Give total number of items :"))
    print("Give weight and value of items :")
    
    items = list()
    for i in range(0, total_items):
        tokens = [int(x) for x in input().split()]
        items.append(item(tokens[0], tokens[1]))
    assert(len(items) == total_items)
    print("The maximum value knapsack can hold is : ",knapsack(items, total_weight, total_items))

if __name__ == "__main__":
    main()

File: Knapsack/Repeated/Knapsack.java
import java.util.Scanner;

/*
    ****************DYNAMIC PROGRAMMING******************
    * This program computes discrete knapsack value with repetitions allowed
    * It stores optimized  discrete knapsack value for each weight 
    * from [1.....W] where W is the total_capacity of the given knapsack.
    * It is same as Coin change problem (Dynamic programming).
    * It takes O(n * W) time.
*/

class item
{
    public int weight;
    public int value;

    public item(int weight, int value)
    {
        this.weight = weight;
        this.value = value;
    }
}

public class Knapsack
{
    private static int knapsack(item items[], int totalItems, int totalWeight)
    {
        int values_of_weights[] = new int[totalWeight + 1];
        values_of_weights[0] = 0;
        int current_value = 0;
        for(int olc = 1; olc <= totalWeight; olc++)
        {
            values_of_weights[olc] = 0;
            for(int ilc = 0; ilc < totalItems; ilc++)
            {
                if(items[ilc].weight <= olc)
                {
                    current_value = values_of_weights[olc - items[ilc].weight] + items[ilc].value;
                }
                if(current_value > values_of_weights[olc])
                {
                    values_of_weights[olc] = current_value;
                }
            }
        }
        return values_of_weights[totalWeight];
    }

    public static void main(String args[])
    {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("Give total capacity of knapsack :");
        int totalWeight = scanner.nextInt();

        System.out.println("Give total number of items :");
        int totalItems = scanner.nextInt();

        System.out.println("Give weights and values of items :");
        item items[] = new item[totalItems];

        for(int lc = 0; lc < totalItems; lc++)
        {
            items[lc] = new item(scanner.nextInt(), scanner.nextInt());
        }

        System.out.println("Maximum value the knapsack can hold is : " + knapsack(items, totalItems, totalWeight));
        scanner.close();
    }
}